win32/tests/threadpool_win32_int/threadpool_win32_int.c

```diff
@@ -952,19 +943,119 @@ static DWORD WINAPI chaos_thread_with_timers_no_lock_func(LPVOID lpThreadParamet
             WakeByAddressSingle((PVOID)&chaos_test_data->timers_starting);
         }
         break;
-        case TEST_ACTION_DESTROY_TIMER:
-            // Stop a timer
+        case TEST_ACTION_CANCEL_TIMER:
+            // Cancel a timer
+        {
+            // Synchronize with close
+            (void)InterlockedIncrement(&chaos_test_data->timers_starting);
+            if (InterlockedAdd(&chaos_test_data->can_start_timers, 0) != 0)
+            {
+                int which_timer_slot = rand() * MAX_TIMER_COUNT / (RAND_MAX + 1);
+                if (InterlockedAdd(&chaos_test_data->timers[which_timer_slot].state, 0) == TIMER_STATE_STARTED)
+                {
+                    threadpool_timer_cancel(chaos_test_data->timers[which_timer_slot].timer);
+                }
+            }
+            (void)InterlockedDecrement(&chaos_test_data->timers_starting);
+            WakeByAddressSingle((PVOID)&chaos_test_data->timers_starting);
+        }
+        break;
+        case TEST_ACTION_RESTART_TIMER:
+            // Restart a timer
+        {
+            // Synchronize with close
+            (void)InterlockedIncrement(&chaos_test_data->timers_starting);
+            if (InterlockedAdd(&chaos_test_data->can_start_timers, 0) != 0)
+            {
+                int which_timer_slot = rand() * MAX_TIMER_COUNT / (RAND_MAX + 1);
+                if (InterlockedAdd(&chaos_test_data->timers[which_timer_slot].state, 0) == TIMER_STATE_STARTED)
+                {
+                    uint32_t timer_start_delay = TIMER_START_DELAY_MIN + rand() * (TIMER_START_DELAY_MAX - TIMER_START_DELAY_MIN) / (RAND_MAX + 1);
+                    uint32_t timer_period = TIMER_PERIOD_MIN + rand() * (TIMER_PERIOD_MAX - TIMER_PERIOD_MIN) / (RAND_MAX + 1);
+                    ASSERT_ARE_EQUAL(int, 0, threadpool_timer_restart(chaos_test_data->timers[which_timer_slot].timer, timer_start_delay, timer_period));
+                }
+            }
+            (void)InterlockedDecrement(&chaos_test_data->timers_starting);
+            WakeByAddressSingle((PVOID)&chaos_test_data->timers_starting);
+        }
+        case TEST_ACTION_SCHEDULE_WORK_ITEM:
+            // perform a schedule work item
+            if (InterlockedAdd(&chaos_test_data->can_schedule_works, 0) != 0)
+            {
+                if (threadpool_schedule_work_item(chaos_test_data->threadpool, chaos_test_data->work_item_context) == 0)
+                {
+                    (void)InterlockedIncrement64(&chaos_test_data->expected_call_count);
+                }
```

Comment:
> } [](http://example.com/codeflow?start=15&length=2)

Consider adding the else branch empty (just with a comment saying that we don't care if it failed.

Reply:
added